### PR TITLE
Add downward-inclined brakes to hybrid coaster and single rail coaster

### PR DESCRIPTION
**Pull request by fidwell**
Adds a gentle-sloped brake as a new track piece. Sprites for this piece are available for the hybrid and single rail coasters, courtesy of X123M3-256.

The brakes are only available in the downward direction. It was suggested on Discord that we also have ones going up, to emulate trim brakes. This are simple to add, but would really only be "useful" for the B&M track. I have started a discussion here #22405 to consider other kinds of inclined brakes.

image
image

Demo .park file with two coasters: [sloped-brakes-demo.zip](https://github.com/user-attachments/files/16385665/sloped-brakes-demo.zip)

### CODE REVIEW

File: src/openrct2-ui/windows/RideConstruction.cpp [hunk 1]
```diff
@@ -1458,6 +1458,7 @@ static Widget _rideConstructionWidgets[] = {
                     break;
                 case TrackElemType::BlockBrakes:
                 case TrackElemType::DiagBlockBrakes:
+                case TrackElemType::SlopedBrakes:
```
Comment (spacek531) [inline src/openrct2-ui/windows/RideConstruction.cpp:1461]: This is only there to gently encourage people to build block brakes with default speed for TD6 compatibility. It's not necessary for brakes.

Comment (fidwell) [reply src/openrct2-ui/windows/RideConstruction.cpp]: Thanks! I saw "brakes" and got excited.

**Review by karst — APPROVED** [review]
This feels very pleasant to play with :)

8bfe6ab1f53cef23c98eeddcd4d84deb 1

**Comment by AaronVanGeffen** [comment]
Could we add @X123M3-256 as co-author to when we merge the PR?

File: src/openrct2/ride/TrackData.cpp [hunk 2]
```diff
@@ -6308,6 +6317,7 @@ static constexpr uint32_t TrackFlags[] = {
     /* TrackElemType::RightEighthBankToOrthogonalDown25      */   TRACK_ELEM_FLAG_TURN_RIGHT | TRACK_ELEM_FLAG_TURN_SLOPED | TRACK_ELEM_FLAG_DOWN | TRACK_ELEM_FLAG_STARTS_AT_HALF_HEIGHT | TRACK_ELEM_FLAG_ALLOW_LIFT_HILL | TRACK_ELEM_FLAG_CURVE_ALLOWS_LIFT,  
     /* TrackElemType::DiagBrakes                                    */   0,
     /* TrackElemType::DiagBlockBrakes                               */   0,
+    /* TrackElemType::Down25                                        */   TRACK_ELEM_FLAG_DOWN | TRACK_ELEM_FLAG_STARTS_AT_HALF_HEIGHT,
```
Comment (spacek531) [inline src/openrct2/ride/TrackData.cpp]: /* TrackElemType::SlopedBrakes

File: src/openrct2/ride/TrackData.cpp [hunk 3]
```diff
@@ -6656,6 +6666,7 @@ static constexpr TrackDefinition TrackDefinitions[] =
     { TRACK_SLOPE_CURVE_LARGE_BANKED,               TrackPitch::Down25, TrackPitch::Down25, TrackRoll::Right,       TrackRoll::Right,    -16 }, // TrackElemType::RightEighthBankToOrthogonalDown25
     { TRACK_DIAG_BRAKES,            TrackPitch::None,           TrackPitch::None,           TrackRoll::None,        TrackRoll::None,        0  }, // TrackElemType::DiagBrakes
     { TRACK_DIAG_BLOCK_BRAKES,      TrackPitch::None,           TrackPitch::None,           TrackRoll::None,        TrackRoll::None,        0  }, // TrackElemType::DiagBlockBrakes
+    { TRACK_FLAT,                   TrackPitch::Down25,        TrackPitch::Down25,        TrackRoll::None,        TrackRoll::None,        0                  },  // TrackElemType::SlopedBrakes
```
Comment (spacek531) [inline src/openrct2/ride/TrackData.cpp]: some whitespace shifting is needed

File: src/openrct2/ride/TrackData.cpp (hunk 2 above)
Comment (fidwell) [reply src/openrct2/ride/TrackData.cpp]: You must have seen an outdated version; this comment has been corrected.

File: src/openrct2/ride/TrackData.cpp (hunk 3 above)
Comment (fidwell) [reply src/openrct2/ride/TrackData.cpp]: This was also outdated code. The whitespace in this block is all a bit wonky, but I tried to make mine match the lines above it.

**Comment by X123M3-256** [comment]
>There is a tiny discrepancy in the support shading compared to the brakeless down-25° piece. I might edit these by hand to match better.

Can you post the comparison? I don't think there should be a discrepancy because IIRC, I disabled AO for the support models, which should be the only thing that could cause the shading to differ. It could be that my renderer is misconfigured or has a bug. I often make changes and don't always check if those changes are breaking older files.

**Comment by fidwell** [comment]
> Can you post the comparison?

Sure; here it is compared to the regular gentle slope piece.
hybrid-compare

If you zoom in, you can see that the wooden support is 6 pixels thick, and the shaded area is about a pixel wider in the new version. It's minor, but it does make the support noticeably darker compared to the original renders.

image

**Comment by X123M3-256** [comment]
> If you zoom in, you can see that the wooden support is 6 pixels thick, and the shaded area is about a pixel wider in the new version. It's minor, but it does make the support noticeably darker compared to the original renders.

It doesn't look like that's the only discrepancy either, the steel beams also look darker. It does look like something isn't right. It looks like the cross ties are darker than they should be too. My first guess is that it is not correctly loading the lighting - the hybrid uses different lighting than the default for my tool.

If you're going to try editing by hand, just cut out the brake sprite and paste it over the original gentle slope sprite. I will see if I can find out what is wrong, because if I can't render sprites with matching shading then no more new pieces can be added (for most new pieces, fixing it by hand is far from practical). Might need to check the commit log and see if I changed anything that could have broken this.

**Comment by fidwell** [comment]
> If you're going to try editing by hand, just cut out the brake sprite and paste it over the original gentle slope sprite.

Yep, that's exactly what I was planning on doing. Thanks for looking into it.

**Comment by Realsteel89** [comment]
> It doesn't look like that's the only discrepancy either, the steel beams also look darker. It does look like something isn't right. It looks like the cross ties are darker than they should be too. My first guess is that it is not correctly loading the lighting - the hybrid uses different lighting than the default for my tool.
> 
> If you're going to try editing by hand, just cut out the brake sprite and paste it over the original gentle slope sprite. I will see if I can find out what is wrong, because if I can't render sprites with matching shading then no more new pieces can be added (for most new pieces, fixing it by hand is far from practical). Might need to check the commit log and see if I changed anything that could have broken this.

if the new sprites were made in windows then they have a difference because windows changes the material a bit. but i already said that half a year ago there are slight differences in the track render of windows and linux
image

**Comment by fidwell** [comment]
All fixed now.

image

**Comment by fidwell** [comment]
I also have implemented upward-inclined brakes, potentially for trim braking. Worth adding to this PR, or no?

image

**Comment by SpartanFrederic104** [comment]
Go for it! I'm also curious if there will be downward-inclined block brakes too.

**Comment by karst** [comment]
I think it's worth adding, tho I don't know any RMC from the top of my head that uses trims.

B&M on the other hand...

**Comment by astraylife1** [comment]
For B&Ms, the trims should be on the steep slope

**Comment by fidwell** [comment]
I made a discussion topic here if you want to discuss other ride types: #22405

**Comment by Realsteel89** [comment]
> I also have implemented upward-inclined brakes, potentially for trim braking. Worth adding to this PR, or no?
> 
> image
> 

Did they worked correktly I can add them to the track code I spitted now the code a bit more in using it in one round with the right brake object.

**Comment by fidwell** [comment]
They work, but I haven't made them public yet. I am hoping to get more feedback from the development team first.

**Comment by Realsteel89** [comment]
Maybe I can use the code for my b&m track adding

**Comment by spacek531** [comment]
Are there any other changes for this PR?

**Comment by fidwell** [comment]
I don't think so. Just needs a rebase.

File: src/openrct2/paint/track/coaster/SingleRailRollerCoaster.cpp [hunk 4]
```diff
@@ -18391,6 +18391,51 @@ namespace OpenRCT2::SingleRailRC
         TrackLeftEighthBankToDiagUp25(session, ride, trackSequence, (direction + 3) & 3, height, trackElement);
     }
 
+    static void Track25DegDownBrakes(
```
Comment (spacek531) [inline src/openrct2/paint/track/coaster/SingleRailRollerCoaster.cpp:18394]: Can you change this function to use the hybrid coaster's Track25DegDownBrakes as the basis? It helps unify the code which will help making track paint into objects.

**Review by spacek531 — CHANGES_REQUESTED** [review]
This is good and straightforward. Only one thing to address with the single rail paint code and it's ready 👍

**Review by spacek531 — APPROVED** [review]
LGTM!

**Comment by Gymnasiast** [comment]
I rebased the PR for you, and added a track group for it. You can’t put it in the regular brakes track group, because that means that the element will be enabled for every ride that supports brakes.

**Comment by Gymnasiast** [comment]
I’ll continue reviewing/testing at another time.

**Comment by karst** [comment]
> Could we add @X123M3-256 as co-author to when we merge the PR?

Can you make sure this is done before merging? 

I will now test it

**Comment by Gymnasiast** [comment]
> > Could we add @X123M3-256 as co-author to when we merge the PR?
> 
> Can you make sure this is done before merging?

I will, thanks for reminding me.

**Review by karst — APPROVED** [review]
After playtesting I can say:
- Not available on rides they shouldn't be
- Brakes function correctly and as expected.
- Nice work! :)
image

**Review by Gymnasiast — APPROVED** [review]

**Comment by tupaschoal** [comment]
Should it not have bumped network version too?

**Comment by Gymnasiast** [comment]
It should have, yes.

**Comment by fidwell** [comment]
I thought it did at one point... probably got lost in a merge somewhere along the line. Sorry about that

**Comment by Gymnasiast** [comment]
Could have been my rebase as well. It’s not a big deal anyway.

**Pull request merged**